File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        opens = []
        if len(s) % 2 != 0:
            return False
        for i in s:
            if i == "(" or i == "{" or i == "[":
                opens.append(i)
            else:
                if len(opens) == 0:
                    return False

                character = opens.pop()
                if (
                    (character == "{" and i != "}")
                    or (character == "[" and i != "]")
                    or (character == "(" and i != ")")
                ):
                    return False

        if len(opens) != 0:
            return False
        return True

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> opens = new Stack<>();
        if (s.length() % 2 != 0)
        return false;
        for (int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if (c == '(' || c == '{' || c == '[')
            opens.push(c);
            else{
                if (opens.size() == 0)
                return false;
                char character = opens.pop();
                if ((character == '{' && c != '}') || (character == '[' && c != ']') || (character == '(' && c != ')'))
                return false;
            }
        }
        if (opens.size() != 0)
        return false;
        return true;
    }
}
